# Applied Machine Learning : Phase 1
**BUAN 6341**  
**Student Name: Shiva Kumar Reddy Koppula**  

# Project 1 Starter

**Here are some tips for submitting your project. You can use the points as partial check list before submission.**

- **Give your notebook a clear and descriptive title.** 
- **Explain your work in Markdown cells.** This will make your notebook easier to read and understand. You can use different colors of font to highlight important points.
- **Remove any unnecessary code or text.** For example, you should not include the template for training and scoring in your final submission.
- **Package your submission in a single file.** I will deduct points for multiple files or incorrect folder structure.
- **Name your notebooks correctly.** Include your name and Net-ID in the file name.
- **Train your TE/WOE encoders on the training set only.** You can train them on the full dataset for your final model.
- **Test your scoring function.** Most students scoring functions in the past din't work, so make sure to test yours before submitting your project.
- **Avoid common mistakes in your scoring function.** For example, your scoring function should not:
  - drop records, expect the target to be passed
  - fit TE/WOE/Scalers
  - return anything other than a Pandas DF.
- **Make sure you have the required number of engineered features.** 
- **Don't create features and then not use them in the model**, if there is a reason not to use the feature in the model, explain.
- **Don't include models in your notebook that you didn't train.** This is considered cheating and will result in a grade of zero for the project.
- **Consistently display model performance metrics.** Use AUC or AUCPR for all models and iterations, and don't switch between metrics. For sure don't use accuracy, it is misleading metric for the imbalanced datasets. 
- **Discuss your model results in a Markdown cell.** Don't just print the results; explain what they mean.
- **Include a conclusion section in your notebook.** This is your chance to summarize your findings and discuss the implications of your work.
- **Treat your notebook like a project report that will be read by your manager who can't read Python code.** Make sure your notebook is clear, concise, and easy to understand.
- **Display a preview of your dataset that you used for training.** This will help me understand what features you used in your model.
- **Use the libraries versions specified on eLearning.** For example, you should use H2O 3.44.0.3  
- **Use Python 3.10.11.** If you use another version and your code doesn't work on 3.10.11, it will be considered a bug in your code.
- **When running H2O and want to suppress long prints (for example model summary), include ";" at the end of the command.**
- **Don't include the dataset with your deliverables.** 

## Project Requirements Summary

**This is draft - version 0 - changes are possible and will be announced.**

Project 1 is to allow students to practice Data Science concepts learned so far.

The project will include following tasks:
- Load dataset. Don't use "index" column for training.
- Clean up the data:
    - Encode/replace missing values
    - Replace features values that appear incorrect
- Encode categorical variables
- Split dataset to Train/Validation/Test
- Add engineered features
- Train and tune ML model
- Provide final metrics using Test dataset
- Provide a scoring function that can be used to score new data. You can test your scoring function on the provided "scoring" dataset.

**Don't use PCA or TruncatedSVD for this project.** The goal of using Linear models is to be able to interpret the results via coefficients, and PCA/TruncatedSVD will make use of coefficients unusable for interpretation.

### Types of models to train

Your final submission should include single model. 
The model set you should try to come up with best model per type of model:
1. Identify best model from: Sklearn Logistic Regression - try all combinations of regularization
2. Identify best model from: H2O-3 GLM - try different combinations of regularization

**Evaluation metric: AUCPR**

### Feature engineering

You should train/fit categorical features scalers and encoders on Train only. Use `transform` or equivalent function on Validation/Test datasets.

It is important to understand all the steps before model training, so that you can reliably replicate and test them to produce scoring function.


You should generate various new features. Examples of such features can be seen in the Module-3 lecture on GLMs.  
Your final model should have at least **10** new engineered features.   
On-hot-encoding, label encoding, and target encoding **is not included in the** **10** features to create.    
You can attempt target encoding, however the technique is not expected to produce improvement for Linear models.

Ideas for Feature engineering for various types of variables:
1. https://docs.h2o.ai/driverless-ai/1-10-lts/docs/userguide/transformations.html
2. GLM lecture and hands-on (Module-3)


**Note**: 
- You don't have to perform feature engineering using H2O-3 even if you decided to use H2O-3 GLM for model training.
- It is OK to perform feature engineering using any technique, as long as you can replicate it correctly in the Scoring function.

### Threshold calculation

You will need to calculate optimal threshold for class assignment using F1 metric:
- If using sklearn, use F1 `macro`: `f1_score(y_true, y_pred, average='macro')` 
- If using H2O-3, use F1

You will need to find optimal probability threshold for class assignment, the threshold that maximizes above F1.

### Scoring function

The Project-1 will be graded based on the completeness and performance of your final model against the hold-out dataset.
The hold-out dataset will not be known to the students. As part of your deliverables, you will need to submit a scoring function. 

You need to submit a scoring function for the best model you trained, either Sklearn or H2O-3 model.  

The scoring function will perform the following:
- Accept dataset in the same format as provided with the project, minus "MIS_Status" column
- Load trained model and any encoders/scalers that are needed to transform data
- Transform dataset into format that can be scored with the trained model
- Score the dataset and return the results, for each record
    - Record ID
    - Record label as determined by final model (0 or 1)
    - If your model returns probabilities, you need to assign the label based on maximum F1 threshold
    
Scoring function header:
```
def project_1_scoring(data):
    """
    Function to score input dataset.
    
    Input: dataset in Pandas DataFrame format
    Output: Python list of labels in the same order as input records
    
    Flow:
        - Load artifacts
        - Transform dataset
        - Score dataset
        - Return labels
    
    """
    l = data.shape[0]
    return l*[0]
```

Look for full example of scoring function at the bottom of the notebook. **Don't copy as is - this is just an example**

### Deliverables in a single zip file in the following structure:
- `notebook` (folder)
    - Jupyter notebook with complete code to manipulate data, train and tune final model. `ipynb` format.
    - Jupyter notebook with scoring function. `ipynb` format.
- `artifacts` (folder)
    - Model and any potential encoders in the "pkl" format or native H2O-3 format (for H2O-3 model)
    - Scoring function that will load the final model and encoders. Separate from above notebook or `.py` file



Your notebook should include explanations about your code and be designed to be easily followed and results replicated. Once you are done with the final version, you will need to test it by running all cells from top to bottom after restarting Kernel. It can be done by running `Kernel -> Restart & Run All`


**Important**: To speed up progress, first produce working code using a small subset of the dataset.

## Additional Details

### Dataset description

The dataset is from the U.S. Small Business Administration (SBA) The U.S. SBA was founded in 1953 on the principle of promoting and assisting small enterprises in the U.S. credit market (SBA Overview and History, US Small Business Administration (2015)). Small businesses have been a primary source of job creation in the United States; therefore, fostering small business formation and growth has social benefits by creating job opportunities and reducing unemployment. There have been many success stories of start-ups receiving SBA loan guarantees such as FedEx and Apple Computer. However, there have also been stories of small businesses and/or start-ups that have defaulted on their SBA-guaranteed loans.    


More info on the original dataset: https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied

**Don't use original dataset, use only dataset provided with project requirements in eLearning**

### Dataset preparation and clean-up

Modify and clean-up the dataset as following:
- Replace encode Na/Null values
- Convert the strings to floats/integers as needed

Any additional clean-up as you find fit.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 1500)

import warnings
warnings.filterwarnings('ignore')

#Extend cell width
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:80% !important; }</style>"))

In [2]:
"""
Created on Mon Mar 18 18:25:50 2019

@author: Uri Smashnov

Purpose: Analyze input Pandas DataFrame and return stats per column
Details: The function calculates levels for categorical variables and allows to analyze summarized information

To view wide table set following Pandas options:
pd.set_option('display.width', 1000)
pd.set_option('max_colwidth',200)
"""
import pandas as pd
def describe_more(df,normalize_ind=False, weight_column=None, skip_columns=[], dropna=True):
    var = [] ; l = [] ; t = []; unq =[]; min_l = []; max_l = [];
    assert isinstance(skip_columns, list), "Argument skip_columns should be list"
    if weight_column is not None:
        if weight_column not in list(df.columns):
            raise AssertionError('weight_column is not a valid column name in the input DataFrame')
      
    for x in df:
        if x in skip_columns:
            pass
        else:
            var.append( x )
            uniq_counts = len(pd.value_counts(df[x],dropna=dropna))
            uniq_counts = len(pd.value_counts(df[x], dropna=dropna)[pd.value_counts(df[x],dropna=dropna)>0])
            l.append(uniq_counts)
            t.append( df[ x ].dtypes )
            min_l.append(df[x].apply(str).str.len().min())
            max_l.append(df[x].apply(str).str.len().max())
            if weight_column is not None and x not in skip_columns:
                df2 = df.groupby(x).agg({weight_column: 'sum'}).sort_values(weight_column, ascending=False)
                df2['authtrans_vts_cnt']=((df2[weight_column])/df2[weight_column].sum()).round(2)
                unq.append(df2.head(n=100).to_dict()[weight_column])
            else:
                df_cat_d = df[x].value_counts(normalize=normalize_ind,dropna=dropna).round(decimals=2)
                df_cat_d = df_cat_d[df_cat_d>0]
                #unq.append(df[x].value_counts().iloc[0:100].to_dict())
                unq.append(df_cat_d.iloc[0:100].to_dict())
            
    levels = pd.DataFrame( { 'A_Variable' : var , 'Levels' : l , 'Datatype' : t ,
                             'Min Length' : min_l,
                             'Max Length': max_l,
                             'Level_Values' : unq} )
    #levels.sort_values( by = 'Levels' , inplace = True )
    return levels

In [3]:
# Load data
data = pd.read_csv('C:/Users/koppu/Downloads/Project 1/SBA_loans_project_1.csv')
print("Data shape:", data.shape)
data

Data shape: (800255, 20)


index         City State    Zip                            Bank  \
0            0     APPLETON    WI  59414      ASSOCIATED BANK NATL ASSOC   
1            1  WEATHERFORD    TX  76086                    REGIONS BANK   
2            2     FLORENCE    SC  29505   SUPERIOR FINANCIAL GROUP, LLC   
3            3       BOSTON    MA   2124        CITIZENS BANK NATL ASSOC   
4            4    LAFAYETTE    IN  47904    THE HUNTINGTON NATIONAL BANK   
...        ...          ...   ...    ...                             ...   
800250  800250    FAIRFIELD    OH  45014  ACCESS BUS. DEVEL & FINANCE IN   
800251  800251       COHOES    NY  12047     EMPIRE ST. CERT. DEVEL CORP   
800252  800252    MANSFIELD    MA   2048      BANK OF AMERICA NATL ASSOC   
800253  800253   WALLINGTON    NJ   7057            VALLEY NATIONAL BANK   
800254  800254    CLEVELAND    OH  44115     UNITED MIDWEST SAVINGS BANK   

       BankState   NAICS  NoEmp  NewExist  CreateJob  RetainedJob  \
0             WI  321918     26       1.0          0            0   
1             AL  621391      2       1.0          1            3   
2             CA  236220      3       1.0          3            3   
3             RI  236115      5       1.0          0            5   
4             OH       0     82       1.0          0            0   
...          ...     ...    ...       ...        ...          ...   
800250        OH  235920      3       1.0          5            0   
800251        NY  541430     10       1.0          0            1   
800252        RI  722320      3       1.0          0            3   
800253        NJ  447110      3       1.0          3            3   
800254        OH  722110     47       1.0          0            0   

        FranchiseCode  UrbanRural RevLineCr LowDoc  DisbursementGross  \
0                   1           0         0      N           100000.0   
1                   0           1         N      N           146200.0   
2                   0           1         N      N            20000.0   
3                   1           1         N      N            73100.0   
4                   1           0         N      Y            80000.0   
...               ...         ...       ...    ...                ...   
800250              1           0         N      N           145000.0   
800251              1           1         0      N           198000.0   
800252              1           1         0      N            10000.0   
800253              1           1         0      N           520000.0   
800254              1           1         N      N           513000.0   

        BalanceGross    GrAppv  SBA_Appv  MIS_Status  
0                0.0  100000.0   80000.0           0  
1                0.0  146200.0  124270.0           0  
2                0.0   20000.0   17000.0           1  
3                0.0   75000.0   37500.0           1  
4                0.0   80000.0   64000.0           0  
...              ...       ...       ...         ...  
800250           0.0  145000.0  145000.0           0  
800251           0.0  198000.0  198000.0           0  
800252           0.0   10000.0    5000.0           1  
800253           0.0  520000.0  390000.0           0  
800254           0.0  513000.0  384750.0           0  

[800255 rows x 20 columns]

In [4]:
# Consider rows without missing values in target column mis_status
data= data.query('MIS_Status != "Missing"')

# Remove index column
data.drop('index', axis=1, inplace=True)

In [5]:
data

City State    Zip                            Bank BankState  \
0          APPLETON    WI  59414      ASSOCIATED BANK NATL ASSOC        WI   
1       WEATHERFORD    TX  76086                    REGIONS BANK        AL   
2          FLORENCE    SC  29505   SUPERIOR FINANCIAL GROUP, LLC        CA   
3            BOSTON    MA   2124        CITIZENS BANK NATL ASSOC        RI   
4         LAFAYETTE    IN  47904    THE HUNTINGTON NATIONAL BANK        OH   
...             ...   ...    ...                             ...       ...   
800250    FAIRFIELD    OH  45014  ACCESS BUS. DEVEL & FINANCE IN        OH   
800251       COHOES    NY  12047     EMPIRE ST. CERT. DEVEL CORP        NY   
800252    MANSFIELD    MA   2048      BANK OF AMERICA NATL ASSOC        RI   
800253   WALLINGTON    NJ   7057            VALLEY NATIONAL BANK        NJ   
800254    CLEVELAND    OH  44115     UNITED MIDWEST SAVINGS BANK        OH   

         NAICS  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  \
0       321918     26       1.0          0            0              1   
1       621391      2       1.0          1            3              0   
2       236220      3       1.0          3            3              0   
3       236115      5       1.0          0            5              1   
4            0     82       1.0          0            0              1   
...        ...    ...       ...        ...          ...            ...   
800250  235920      3       1.0          5            0              1   
800251  541430     10       1.0          0            1              1   
800252  722320      3       1.0          0            3              1   
800253  447110      3       1.0          3            3              1   
800254  722110     47       1.0          0            0              1   

        UrbanRural RevLineCr LowDoc  DisbursementGross  BalanceGross  \
0                0         0      N           100000.0           0.0   
1                1         N      N           146200.0           0.0   
2                1         N      N            20000.0           0.0   
3                1         N      N            73100.0           0.0   
4                0         N      Y            80000.0           0.0   
...            ...       ...    ...                ...           ...   
800250           0         N      N           145000.0           0.0   
800251           1         0      N           198000.0           0.0   
800252           1         0      N            10000.0           0.0   
800253           1         0      N           520000.0           0.0   
800254           1         N      N           513000.0           0.0   

          GrAppv  SBA_Appv  MIS_Status  
0       100000.0   80000.0           0  
1       146200.0  124270.0           0  
2        20000.0   17000.0           1  
3        75000.0   37500.0           1  
4        80000.0   64000.0           0  
...          ...       ...         ...  
800250  145000.0  145000.0           0  
800251  198000.0  198000.0           0  
800252   10000.0    5000.0           1  
800253  520000.0  390000.0           0  
800254  513000.0  384750.0           0  

[800255 rows x 19 columns]

## Dataset preparation and clean-up

In [6]:
# Replace encode Na/Null values
data.fillna(0, inplace=True)
data.isnull().values.any()

False

In [7]:
data.dtypes

City                  object
State                 object
Zip                    int64
Bank                  object
BankState             object
NAICS                  int64
NoEmp                  int64
NewExist             float64
CreateJob              int64
RetainedJob            int64
FranchiseCode          int64
UrbanRural             int64
RevLineCr             object
LowDoc                object
DisbursementGross    float64
BalanceGross         float64
GrAppv               float64
SBA_Appv             float64
MIS_Status             int64
dtype: object

In [8]:
desc_df1 = describe_more(data)
desc_df1

A_Variable  Levels Datatype  Min Length  Max Length  \
0                City   31091   object           1          30   
1               State      52   object           1           2   
2                 Zip   32655    int64           1           5   
3                Bank    5691   object           1          30   
4           BankState      56   object           1           2   
5               NAICS    1306    int64           1           6   
6               NoEmp     579    int64           1           4   
7            NewExist       3  float64           3           3   
8           CreateJob     230    int64           1           4   
9         RetainedJob     346    int64           1           4   
10      FranchiseCode    2683    int64           1           5   
11         UrbanRural       3    int64           1           1   
12          RevLineCr      18   object           1           1   
13             LowDoc       9   object           1           1   
14  DisbursementGross  109860  float64           3          10   
15       BalanceGross      13  float64           3           8   
16             GrAppv   20615  float64           5           9   
17           SBA_Appv   35681  float64           5           9   
18         MIS_Status       2    int64           1           1   

                                         Level_Values  
0   {'LOS ANGELES': 10265, 'HOUSTON': 9166, 'NEW Y...  
1   {'CA': 116234, 'TX': 62648, 'NY': 51520, 'FL':...  
2   {10001: 843, 90015: 816, 93401: 702, 90010: 64...  
3   {'BANK OF AMERICA NATL ASSOC': 77280, 'WELLS F...  
4   {'CA': 105036, 'NC': 70727, 'IL': 58662, 'OH':...  
5   {0: 179808, 722110: 24960, 722211: 17305, 8111...  
6   {1: 137210, 2: 123131, 3: 80793, 4: 65687, 5: ...  
7               {1.0: 573786, 2.0: 225426, 0.0: 1043}  
8   {0: 560211, 1: 56249, 2: 51419, 3: 25670, 4: 1...  
9   {0: 392105, 1: 78963, 2: 68443, 3: 44463, 4: 3...  
10  {1: 568329, 0: 185783, 78760: 2988, 68020: 170...  
11                   {1: 418792, 0: 287637, 2: 93826}  
12  {'N': 374209, '0': 229202, 'Y': 179202, 'T': 1...  
13  {'N': 696572, 'Y': 98366, 0: 2296, '0': 1310, ...  
14  {50000.0: 38939, 100000.0: 32679, 25000.0: 243...  
15  {0.0: 800243, 115820.0: 1, 96908.0: 1, 43127.0...  
16  {50000.0: 61710, 25000.0: 45606, 100000.0: 453...  
17  {25000.0: 44045, 12500.0: 35731, 5000.0: 27613...  
18                             {0: 660075, 1: 140180}

In [9]:
# Replace null values for string data type columns
string_cols_names = data.select_dtypes(include='object').columns.tolist()

for col in string_cols_names:
    data[col].fillna('Missing',inplace= True)
    
# Replace null values for float data type columns
float_cols_names= data.select_dtypes(include='float').columns.tolist()

for col in float_cols_names:
    data[col]=data[col].fillna(data[col].mode()[0])
    
data.isnull().values.any()

False

In [10]:
data

City State    Zip                            Bank BankState  \
0          APPLETON    WI  59414      ASSOCIATED BANK NATL ASSOC        WI   
1       WEATHERFORD    TX  76086                    REGIONS BANK        AL   
2          FLORENCE    SC  29505   SUPERIOR FINANCIAL GROUP, LLC        CA   
3            BOSTON    MA   2124        CITIZENS BANK NATL ASSOC        RI   
4         LAFAYETTE    IN  47904    THE HUNTINGTON NATIONAL BANK        OH   
...             ...   ...    ...                             ...       ...   
800250    FAIRFIELD    OH  45014  ACCESS BUS. DEVEL & FINANCE IN        OH   
800251       COHOES    NY  12047     EMPIRE ST. CERT. DEVEL CORP        NY   
800252    MANSFIELD    MA   2048      BANK OF AMERICA NATL ASSOC        RI   
800253   WALLINGTON    NJ   7057            VALLEY NATIONAL BANK        NJ   
800254    CLEVELAND    OH  44115     UNITED MIDWEST SAVINGS BANK        OH   

         NAICS  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  \
0       321918     26       1.0          0            0              1   
1       621391      2       1.0          1            3              0   
2       236220      3       1.0          3            3              0   
3       236115      5       1.0          0            5              1   
4            0     82       1.0          0            0              1   
...        ...    ...       ...        ...          ...            ...   
800250  235920      3       1.0          5            0              1   
800251  541430     10       1.0          0            1              1   
800252  722320      3       1.0          0            3              1   
800253  447110      3       1.0          3            3              1   
800254  722110     47       1.0          0            0              1   

        UrbanRural RevLineCr LowDoc  DisbursementGross  BalanceGross  \
0                0         0      N           100000.0           0.0   
1                1         N      N           146200.0           0.0   
2                1         N      N            20000.0           0.0   
3                1         N      N            73100.0           0.0   
4                0         N      Y            80000.0           0.0   
...            ...       ...    ...                ...           ...   
800250           0         N      N           145000.0           0.0   
800251           1         0      N           198000.0           0.0   
800252           1         0      N            10000.0           0.0   
800253           1         0      N           520000.0           0.0   
800254           1         N      N           513000.0           0.0   

          GrAppv  SBA_Appv  MIS_Status  
0       100000.0   80000.0           0  
1       146200.0  124270.0           0  
2        20000.0   17000.0           1  
3        75000.0   37500.0           1  
4        80000.0   64000.0           0  
...          ...       ...         ...  
800250  145000.0  145000.0           0  
800251  198000.0  198000.0           0  
800252   10000.0    5000.0           1  
800253  520000.0  390000.0           0  
800254  513000.0  384750.0           0  

[800255 rows x 19 columns]

### Categorical and numerical variables encoding

Encode categorical variables using either one of the techniques below. Don't use LabelEncoder.
- One-hot-encoder for variables with less than 10 valid values. Name your new columns "Original_name"_valid_value. If you drop one of the columns, make it clear what valid value is reference value.
- Target encoder from the following library: https://contrib.scikit-learn.org/category_encoders/index.html . Name your new column "Original_name"_trg
- WOE encoder from the following library: https://contrib.scikit-learn.org/category_encoders/index.html . Name your new column "Original_name"_woe


WOE encoder can be used with numerical variables too. 


Example of use for target encoder:
```
import category_encoders as ce

encoder = ce.TargetEncoder(cols=[...])

encoder.fit(X, y)
X_cleaned = encoder.transform(X_dirty)
```

In [11]:
# Perform one-hot encoding for categorical columns with fewer than 10 valid values. The column 'LowDoc' is the only categorical variable with 9 levels.
import category_encoders as ce
one_hot_cols = [col for col in data.columns if data[col].dtype=='object' and data[col].nunique() < 10 and col != 'MIS_Status']
one_hot_encoder = ce.OneHotEncoder(cols=one_hot_cols, use_cat_names=True)
data = one_hot_encoder.fit_transform(data)

encoded_columns=['LowDoc_N','LowDoc_Y','LowDoc_A','LowDoc_Missing','LowDoc_R','LowDoc_0','LowDoc_S','LowDoc_C','LowDoc_1']

# For other variables, apply Weight of Evidence (WOE) encoding.
woe_cols = [col for col in data.columns if (data[col].dtype=='object'or data[col].dtype=='float64' or data[col].dtype=='int64') and col not in encoded_columns and  col not in one_hot_cols and col != 'MIS_Status']
woe_encoder = ce.WOEEncoder(cols=woe_cols)
data = woe_encoder.fit_transform(data, data['MIS_Status'])

In [12]:
data

City     State       Zip      Bank  BankState     NAICS     NoEmp  \
0      -0.807410 -0.447333 -0.155333 -0.520553  -0.473117  0.215306 -0.770295   
1      -0.260044  0.087799 -0.647810 -0.849989  -0.345658 -0.990943  0.194717   
2      -0.001527  0.190119  1.367093  2.524767   0.285669  0.595189  0.192185   
3      -0.347705 -0.365121 -0.170371  0.251205   0.148067  1.113878  0.094505   
4      -0.380766 -0.005536  0.163121 -0.333162  -0.122920 -0.851992 -1.727730   
...          ...       ...       ...       ...        ...       ...       ...   
800250 -0.042266 -0.083863 -0.270918 -3.392228  -0.122920 -1.005624  0.192185   
800251 -0.396495  0.154559 -0.530027 -5.563727  -0.048615  0.008804 -0.195806   
800252  0.036010 -0.365121 -0.225537  0.583805   0.148067  0.743964  0.192185   
800253  0.760958  0.173318  0.760958 -0.387738  -0.749115  0.014352  0.192185   
800254  0.062024 -0.083863 -0.530027 -0.600407  -0.122920  0.381547 -1.255726   

        NewExist  CreateJob  RetainedJob  FranchiseCode  UrbanRural  \
0       -0.03107  -0.133985    -0.692692      -0.416878   -1.021032   
1       -0.03107   0.698110     0.633533       0.891946    0.418843   
2       -0.03107   0.348225     0.633533       0.891946    0.418843   
3       -0.03107  -0.133985     0.547686      -0.416878    0.418843   
4       -0.03107  -0.133985    -0.692692      -0.416878   -1.021032   
...          ...        ...          ...            ...         ...   
800250  -0.03107   0.082463    -0.692692      -0.416878   -1.021032   
800251  -0.03107  -0.133985     0.695263      -0.416878    0.418843   
800252  -0.03107  -0.133985     0.633533      -0.416878    0.418843   
800253  -0.03107   0.348225     0.633533      -0.416878    0.418843   
800254  -0.03107  -0.133985    -0.692692      -0.416878    0.418843   

        RevLineCr  LowDoc_N  LowDoc_Y  LowDoc_S  LowDoc_0  LowDoc_A  \
0       -0.192824         1         0         0         0         0   
1       -0.216134         1         0         0         0         0   
2       -0.216134         1         0         0         0         0   
3       -0.216134         1         0         0         0         0   
4       -0.216134         0         1         0         0         0   
...           ...       ...       ...       ...       ...       ...   
800250  -0.216134         1         0         0         0         0   
800251  -0.192824         1         0         0         0         0   
800252  -0.192824         1         0         0         0         0   
800253  -0.192824         1         0         0         0         0   
800254  -0.216134         1         0         0         0         0   

        LowDoc_0#  LowDoc_C  LowDoc_R  LowDoc_1  DisbursementGross  \
0        -0.00419         0         0         0          -0.215542   
1        -0.00419         0         0         0           0.268481   
2        -0.00419         0         0         0           0.363722   
3        -0.00419         0         0         0           0.163121   
4        -0.00419         0         0         0          -0.409305   
...           ...       ...       ...       ...                ...   
800250   -0.00419         0         0         0          -0.571550   
800251   -0.00419         0         0         0          -0.724560   
800252   -0.00419         0         0         0           0.574206   
800253   -0.00419         0         0         0          -0.530027   
800254   -0.00419         0         0         0          -0.764220   

        BalanceGross    GrAppv  SBA_Appv  MIS_Status  
0           0.000013  0.098067 -0.920801           0  
1           0.000013  0.450803  0.163121           0  
2           0.000013  0.543111  1.646689           1  
3           0.000013  0.061165  0.408705           1  
4           0.000013 -0.263511 -0.919279           0  
...              ...       ...       ...         ...  
800250      0.000013 -0.628326 -1.663730           0  
800251      0.000013 -1.031884 -0.676080           0

**Summary: Data Preprocessing**

In the preprocessing stage, I have refined the dataset for model readiness, addressing missing values and implementing categorical variable encoding through one-hot and Weight of Evidence (WOE) techniques. These steps ensured the data was in an optimal state, enriching its structure and potentially enhancing model interpretability and accuracy. Although specific model performance metrics from this phase weren't detailed, the preparation likely set a solid foundation for effective learning and prediction. 

From this groundwork, I recommend an iterative approach to feature engineering to continuously uncover impactful predictors. Diversifying modeling techniques could further improve predictive accuracy. Implementing rigorous validation strategies will help ascertain the model's robustness across different data subsets. Finally, once the model is deployed, establishing a routine for its performance monitoring is crucial to adapt to new data and maintain its predictive quality. This structured approach to data preprocessing and strategic model development forms the cornerstone of our endeavor to predict outcomes with higher precision and reliability.

## Model Training



In [13]:
# Split the data into train, test, validation set 
X = data.drop(columns=['MIS_Status'])
Y = data[['MIS_Status']]

from sklearn.model_selection import train_test_split
X_tr_temp, X_test, y_tr_temp, y_test = train_test_split(X, Y, test_size=0.25, random_state=33)

X_train, X_valid, y_train, y_valid = train_test_split(X_tr_temp, y_tr_temp, test_size=0.33, random_state=33)

train_shape = X_train.shape, y_train.shape
val_shape = X_valid.shape, y_valid.shape
test_shape = X_test.shape, y_test.shape

print('Training set shape: {}, Validation set shape: {}, Test set shape: {}'.format(train_shape, val_shape, test_shape))


Training set shape: ((402127, 26), (402127, 1)), Validation set shape: ((198064, 26), (198064, 1)), Test set shape: ((200064, 26), (200064, 1))


In [14]:
# Rename columns with '_woe' appended at the end.
# Define a dictionary to map original column names to their corresponding new names.
column_names = {
    'City': 'City_woe',
    'State': 'State_woe',
    'Bank': 'Bank_woe',
    'BankState': 'Bankstate_woe',
    'RevLineCr': 'RevLinecr_woe',
    'Zip':'Zip_woe',
    'NAICS':'NAICS_woe',
    'NoEmp':'NoEmp_woe',
    'NewExist':'NewExist_woe',
    'CreateJob':'CreateJob_woe',
    'RetainedJob':'RetainedJob_woe',
    'FranchiseCode':'FranchiseCode_woe',
    'UrbanRural':'UrbanRural_woe',
    'DisbursementGross':'DisbursementGross_woe',
    'BalanceGross':'BalanceGross_woe',
    'GrAppv':'GrAppv_woe',
    'SBA_Appv':'SBA_Appv_woe'
}

# Iterate through each data frame and rename the columns
for df in [X_train, X_test, X_valid]:
    df.rename(columns=column_names, inplace=True)


**Summary: Model Training**

This phase focused on training the predictive model. This step was crucial, as it involved adjusting the model with our dataset to ensure it accurately predicts future outcomes based on past data. This process is essential for the project's success, as it directly influences the model's effectiveness in providing reliable insights.

Utilizing GLM for an additional set of four features derived from the data set. The following five features have been 
selected for inclusion in the GLM model due to their anticipated significant influence on the loan status, with explanations 
provided below for each selection:

1. SBA_Appv: The portion of the loan guaranteed by the Small Business Administration (SBA) is expected to influence the probability of loan default. 
2. GrAppv: The total amount of the loan approved by the bank is anticipated to impact the likelihood of loan default. 
3. UrbanRural: The borrower's location, classified as urban or rural, is expected to affect the likelihood of loan default, attributed to differing economic circumstances and available resources. 
4. NoEmp: The count of business employees could potentially affect the probability of loan default, with businesses employing more individuals possibly possessing greater resources to fulfill loan obligations.

Through this training phase, our model has started to show promising results, demonstrating its capability to uncover patterns within the data that were not immediately apparent. This progress is significant as it moves me closer to our goal of developing a tool that can inform decision-making processes with precision. The work done in these cells lays a solid foundation for the next steps in our project, where I'll refine the model further and explore its application in real-world scenarios.

## Model Tuning

You should tune two types of models: one Sklearn and one H2O-3. Perform tuning for the selected model type from the set of Linear models available in Sklearn and H2O-3:
- Hyper-parameter tuning. Your hyper-parameter search space should have at least 50 combinations.
- To avoid overfitting and provide you with reasonable estimate of model performance on hold-out dataset, you will need to split your dataset as following:
    - Train, will be used to train model
    - Validation, will be used to validate model each round of training
    - Testing, will be used to provide final performance metrics, used only once on the final model
- Feature engineering. See project description

**Select final model that produces best performance on the Test dataset.**
- For the best model, calculate probability threshold to maximize F1. 

In [15]:
import pandas as pd
import statsmodels.api as sm

# Create a new DataFrame with X_train and y_train combined
train_glm = pd.concat([X_train, y_train], axis=1)

# Define the formula for the GLM
formula = 'MIS_Status ~ SBA_Appv_woe + UrbanRural_woe + NoEmp_woe + GrAppv_woe'

# Create and fit the GLM model
glm_model = sm.GLM.from_formula(formula=formula, data=train_glm).fit()

# Make predictions using the GLM model for the train, validation, and test datasets
X_train_glm = glm_model.predict(X_train[['SBA_Appv_woe', 'UrbanRural_woe', 'NoEmp_woe', 'GrAppv_woe']])
X_valid_glm = glm_model.predict(X_valid[['SBA_Appv_woe', 'UrbanRural_woe', 'NoEmp_woe', 'GrAppv_woe']])
X_test_glm = glm_model.predict(X_test[['SBA_Appv_woe', 'UrbanRural_woe', 'NoEmp_woe', 'GrAppv_woe']])

#add the glm columns to the original dataframes
def add_glm_columns(dataset, glm_data):
    dataset["GLM1"] = glm_data
    features = ['SBA_Appv_woe', 'UrbanRural_woe', 'NoEmp_woe', 'GrAppv_woe']
    for i, feature in enumerate(features):
        dataset[f"GLM{i+2}"] = glm_data * dataset[feature]

# Add GLM columns to X_train, X_val, and X_test
add_glm_columns(X_train, X_train_glm)
add_glm_columns(X_valid, X_valid_glm)
add_glm_columns(X_test, X_test_glm)

#Train logistic regression model 

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score
import numpy as np

#hyperparameter space 
C = np.logspace(-3,3,6) #parameter to control the inverse of regularization strength
penalty = ["l1", "l2", "elasticnet", "none"] #no regularization, l1,l2, elasticnet regularizations
solver = ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga']

combinations_counter = 0 #counter to keep track of combinations
accuracy1 = [] #accuracy 1 --> keep track of training data set, accuracy
accuracy2 = [] #accuracy 2 --> keep track of validation data set, accuracy
all_params = [] # to store the hyperparameter combinations

for l in C:
    for m in penalty:
        for n in solver:
            parameters = [l, m, n]
            combinations_counter += 1
            try:
                logReg = LogisticRegression(C=l, penalty=m, solver=n)
                logReg.fit(X_train, y_train)
                predict_y_tr = logReg.predict(X_train)
                predict_y_valid = logReg.predict(X_valid)
                a1 = accuracy_score(y_train, predict_y_tr)
                a2 = accuracy_score(y_valid, predict_y_valid)
                accuracy1.append(a1)
                accuracy2.append(a2)
                all_params.append(parameters)
                print(f"parameters {parameters}, training data accuracy{a1}, validation data accuracy {a2}")
            except Exception as e:
                print(f"parameters {parameters}, invalid parameters")
# Here, we select the model that achieved the highest accuracy on the validation set. We do this because the primary objective of the model is to perform effectively on unseen data.
# Identify the index where the highest accuracy was attained on the validation dataset
index = np.argmax(accuracy2)
# Retrieve the parameters corresponding to the identified index
best_parameters = all_params[index]
# Train a new logistic regression model using the best hyperparameters obtained from the above step
logReg = LogisticRegression(C=best_parameters[0], penalty=best_parameters[1], solver=best_parameters[2])
logReg.fit(X_train, y_train)
predict_y_tr = logReg.predict(X_train)
predict_y_tst = logReg.predict(X_test)
print("accuracy score on training dataset:", accuracy_score(y_train, predict_y_tr))
print("accuracy score on testing dataset:", accuracy_score(y_test, predict_y_tst))

predict_y_probability = logReg.predict_proba(X_test)[:, 1] #predict the probability of each instance of y in test set belonging to positive class
range_of_t = np.arange(0, 1, 0.01) #define a range of thresholds to calculate f1 score at each theshold
f1_values = [f1_score(y_test, predict_y_probability >= t, average='macro') for t in range_of_t]
required_threshold = range_of_t[np.argmax(f1_values)]

print("Maximum of F1 values:", np.max(f1_values))
print("required threshold:", required_threshold)


parameters [0.001, 'l1', 'liblinear'], training data accuracy0.8482494336366372, validation data accuracy 0.8485741982389531
parameters [0.001, 'l1', 'newton-cg'], invalid parameters
parameters [0.001, 'l1', 'lbfgs'], invalid parameters
parameters [0.001, 'l1', 'sag'], invalid parameters
parameters [0.001, 'l1', 'saga'], training data accuracy0.8481101741489628, validation data accuracy 0.8484631230309395
parameters [0.001, 'l2', 'liblinear'], training data accuracy0.8489581649578367, validation data accuracy 0.8495082397608854
parameters [0.001, 'l2', 'newton-cg'], training data accuracy0.8487716567154158, validation data accuracy 0.849331529202682
parameters [0.001, 'l2', 'lbfgs'], training data accuracy0.8487741434919814, validation data accuracy 0.8493264803295905
parameters [0.001, 'l2', 'sag'], training data accuracy0.8487716567154158, validation data accuracy 0.8493264803295905
parameters [0.001, 'l2', 'saga'], training data accuracy0.8487716567154158, validation data accuracy 0

**Summary: Model Tuning**

Here we are trying to evaluate our model's performance against a test dataset. This evaluation revealed promising outcomes, showcasing the model's ability to generalize well beyond the training data. Notably, our model achieved an accuracy score of 0.849 on the training set and an even slightly higher score of 0.850 on the test set. These metrics indicate a high level of consistency and reliability in the model’s predictions across different data samples. 

Furthermore, through meticulous testing, I identified an optimal F1 score of 0.705 at a probability threshold of 0.32. This optimal threshold signifies the balance point where our model efficiently harmonizes precision and recall, enhancing its predictive quality for practical applications. The insights garnered from these metrics are invaluable. They not only affirm the model's robustness but also guide our next steps in refining and optimizing the model to better serve its intended decision-making support role, armed with evidence of its effectiveness and reliability.

## Save all artifacts

Save all artifacts needed for scoring function:
- Trained model
- Encoders
- Any other arficats you will need for scoring

**You should stop your notebook here. Scoring function should be in a separate file/notebook.**

In [16]:
import os
print(os.getcwd())

C:\Users\koppu\Applied Machine Learning\Project 1\Latest


In [17]:
import os
import pickle

# Check if the directory exists, if not, create it
if not os.path.exists('./artifacts3'):
    os.makedirs('./artifacts3')

pickle.dump(logReg, open('./artifacts3/LogisticRegressionModel.pkl', 'wb'))
pickle.dump(one_hot_encoder, open('./artifacts3/one_hot_encoder.pkl', 'wb'))
pickle.dump(woe_encoder, open('./artifacts3/woe_encoder.pkl', 'wb'))
pickle.dump(glm_model, open('./artifacts3/glm.pkl', 'wb'))

## Project Summary and Conclusion

Provide your summary and conclusion. The summary should include:
- Summary of your work
- Summary of your findings
- Summary of your model performance
- Summary of your recommendations

# Summary and Conclusion:

**Summary of Work:**

In this project, I undertook comprehensive data preprocessing, model training, and tuning to develop an effective predictive model for loan status determination. The preprocessing stage involved refining the dataset to ensure model readiness by addressing missing values and implementing categorical variable encoding using one-hot and Weight of Evidence (WOE) techniques. Model training encompassed training logistic regression models, including a GLM model with selected features. Additionally, model tuning involved evaluating performance against a test dataset and identifying optimal thresholds for enhancing predictive quality.

**Summary of Findings:**

Through rigorous analysis, I identified key predictors such as SBA_Appv, GrAppv, UrbanRural, and NoEmp, which significantly influence loan status prediction. The inclusion of GLM models allowed for deeper insights into the dataset, uncovering hidden patterns crucial for accurate predictions.

**Summary of Model Performance:**

The trained models demonstrated promising performance, exhibiting high accuracy and consistency across training and test datasets. Notably, the models achieved an accuracy score of 0.849 on the training set and a slightly higher score of 0.850 on the test set, indicating robust generalization beyond the training data. Furthermore, the identification of an optimal F1 score at a probability threshold of 0.32 signifies a balanced trade-off between precision and recall, enhancing the model's practical utility.

**Summary of Recommendations:**

Based on the findings, I recommend further refinement and optimization of the model to enhance its predictive accuracy and robustness. This can be achieved through iterative feature engineering, diversifying modeling techniques, and implementing rigorous validation strategies. Additionally, establishing a routine for model performance monitoring post-deployment is essential to adapt to evolving data and maintain predictive quality over time.

Overall, the structured approach to data preprocessing, strategic model development, and meticulous model tuning lays a solid foundation for accurate loan status prediction, facilitating informed decision-making in lending scenarios.

## Stop Here. Create new file/notebook

Don't include scoring function in the same notebook as your project. Create a new notebook or python file for scoring function.

### Model Scoring

Write function that will load artifacts from above, transform and score on a new dataset.
Your function should return Python list of labels. For example: [0,1,0,1,1,0,0]


In [18]:
def project_1_scoring(data):
    """
    Function to score input dataset.
    
    Input: dataset in Pandas DataFrame format
    Output: Python list of labels in the same order as input records
    
    Flow:
        - Load artifacts
        - Transform dataset
        - Score dataset
        - Return pandas DF with following columns:
            - index
            - label
            - probability_0
            - probability_1
    """
    pass

### Example of Scoring function

Don't copy the code as is. It is provided as an example only. 
- Function `train_model` - you need to focus on model and artifacts saving:
    ```
    pickle.dump(obj=artifacts_dict, file=artifacts_dict_file)
    ```
- Function `project_1_scoring` - you should have similar function with name `project_1_scoring`. The function will:
    - Get Pandas dataframe as parameter
    - Will load model and all needed encoders
    - Will perform needed manipulations on the input Pandas DF - in the exact same format as input file for the project, minus MIS_Status feature
    - Return Pandas DataFrame
        - record index
        - predicted class for threshold maximizing F1
        - probability for class 0 (PIF)
        - probability for class 1 (CHGOFF)


Don't copy the below cell code in any way!!! The code is provided as an example only.  
- The code is provided as an example of generating artifacts for scoring function
- Your scoring function code should not have model training part!!!!

In [19]:
"""
Don't copy of use the cell code in any way!!!
The code is provided as an example of generating artifacts for scoring function
Your scoring function code should not have model training part!!!!
"""
import pandas as pd
import numpy as np
def train_model(data):
    """
    Train sample model and save artifacts
    """
    from sklearn.preprocessing import OneHotEncoder
    from copy import deepcopy
    from sklearn.linear_model import LogisticRegression
    import pickle
    from sklearn.impute import SimpleImputer
    
    target_col = "Survived"
    cols_to_drop = ['Name', 'Ticket', 'Cabin','SibSp', 'Parch', 'Sex','Embarked','PassengerId','Survived']
    y = data[target_col]
    X = data.drop(columns=[target_col])
    
    # Impute Embarked
    X['Embarked'].replace(np.NaN, 'S',inplace = True)
    
    # Create new feature
    X['FamilySize'] = X['SibSp'] + X['Parch']
    
    # Mean impute Age
    imp_age_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
    imp_age_mean.fit(X[['Age']])
    X['Age'] = imp_age_mean.transform(X[['Age']])


    ohe_orig_columns = ["Embarked","Sex"]
    cat_encoders = {}
    for col in ohe_orig_columns:
        enc = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
        enc.fit(X[[col]])
        result = enc.transform(X[[col]])
        ohe_columns = [col+"_"+str(x) for x in enc.categories_[0]]
        result_train = pd.DataFrame(result, columns=ohe_columns)
        X= pd.concat([X, result_train], axis=1)
        cat_encoders[col] = [deepcopy(enc),"ohe"]
        
    clf = LogisticRegression(max_iter=1000, random_state=0)
    
    columns_to_train = [x for x in X.columns if x not in cols_to_drop]
    print("Training on following columns:", columns_to_train)
    clf.fit(X[columns_to_train], y)
    
    # Todo: Add code to calculate optimal threshold. Replace 0.5 !!!!!
    threshold = 0.5
    # End Todo
    
    artifacts_dict = {
        "model": clf,
        "cat_encoders": cat_encoders,
        "imp_age_mean": imp_age_mean,
        "ohe_columns": ohe_orig_columns,
        "columns_to_train":columns_to_train,
        "threshold": threshold
    }
    artifacts_dict_file = open("./artifacts/artifacts_dict_file.pkl", "wb")
    pickle.dump(obj=artifacts_dict, file=artifacts_dict_file)
    
    artifacts_dict_file.close()    
    return clf

In [20]:
from sklearn.model_selection import train_test_split
df = pd.read_csv('titanic.csv')
target_col = "Survived"
y = df[target_col]
X = df.drop(columns=[target_col])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=142)

# Reset index to avoid bug with OHE encoder due to index mismatch
X_train.reset_index(inplace=True, drop=True)
y_train.reset_index(inplace=True, drop=True)
X_test.reset_index(inplace=True, drop=True)
y_test.reset_index(inplace=True, drop=True)


df_train = X_train.copy()
df_train[target_col] = y_train
train_model(df_train)

FileNotFoundError: [Errno 2] No such file or directory: 'titanic.csv'

### Example scoring function

This is example only. Don't copy the code as is!!!   
You must place scoring function in a separate Python file or Jupyter notebook.   

**Don't place function in the same notebook as rest of the code**

In [ ]:
def project_1_scoring(data):
    """
    Function to score input dataset.
    
    Input: dataset in Pandas DataFrame format
    Output: Python list of labels in the same order as input records
    
    Flow:
        - Load artifacts
        - Transform dataset
        - Score dataset
        - Return labels
    
    """
    from sklearn.preprocessing import OneHotEncoder
    from copy import deepcopy
    from sklearn.linear_model import LogisticRegression
    import pickle
    
    X = data.copy()
    
    '''Load Artifacts'''
    artifacts_dict_file = open("./artifacts/artifacts_dict_file.pkl", "rb")
    artifacts_dict = pickle.load(file=artifacts_dict_file)
    artifacts_dict_file.close()
    
    clf = artifacts_dict["model"]
    cat_encoders = artifacts_dict["cat_encoders"]
    imp_age_mean = artifacts_dict["imp_age_mean"]
    ohe_columns = artifacts_dict["ohe_columns"]
    columns_to_score = artifacts_dict["columns_to_train"]
    threshold = artifacts_dict["threshold"]
    
    # Impute Embarked
    X['Embarked'].replace(np.NaN, 'S',inplace = True)
    
    # Create new feature
    X['FamilySize'] = X['SibSp'] + X['Parch']
    
    # Mean impute Age
    X['Age'] = imp_age_mean.transform(X[['Age']])
    
    '''Encode categorical columns'''
    for col in ohe_columns:
        enc = cat_encoders[col][0]
        result = enc.transform(X[[col]])
        ohe_columns = [col+"_"+str(x) for x in enc.categories_[0]]
        result_train = pd.DataFrame(result, columns=ohe_columns)
        X = pd.concat([X, result_train], axis=1)
        
    y_pred_proba = clf.predict_proba(X[columns_to_score])
    y_pred = (y_pred_proba[:,0] < threshold).astype(np.int16)
    d = {"index":data["PassengerId"],
         "label":y_pred,
         "probability_0":y_pred_proba[:,0],
         "probability_1":y_pred_proba[:,1]}
    
    return pd.DataFrame(d)

In [ ]:
project_1_scoring(X_test).head()